### PR TITLE
Make blocks registered through registerBlockSingleProductTemplate available in patterns

### PR DESCRIPTION
**Pull request by kmanijak**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Created in collaboration with @gigitux 🙏 

For some time already blocks registered through `registerBlockSingleProductTemplate` were not available in the Patterns due to error in the function logic. When in patterns, these blocks were never initially registered.

For Single Product block it occurred as "Product not registered" error and was raised by a user in https://github.com/woocommerce/woocommerce/issues/51138 after Product Price was migrated to use the `registerBlockSingleProductTemplate` function https://github.com/woocommerce/woocommerce/pull/49906.

Recently I exposed Product Collection collections to the inserter in https://github.com/woocommerce/woocommerce/pull/52209. They use a `createBlocksFromInnerBlocksTemplate` method which internally has additional check for blocks and it throws an error when the block is not registered. After this, when opening an inserter, the Editor crashed. So even though the issue was there for some time (and even raised by a user) already it was "exposed" as destructive by https://github.com/woocommerce/woocommerce/pull/52209.

Closes https://github.com/woocommerce/woocommerce/issues/52699
Closes https://github.com/woocommerce/woocommerce/issues/51138
Closes https://github.com/woocommerce/woocommerce/issues/46781

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Blocks registered by `registerBlockSingleProductTemplate` should follow this logic:

1. Be available "globally" in Single Product template
2. Follow `ancestors` in
    - other templates (e.g. Product Catalog)
    - posts/pages
    - patterns

#### Single Product template

1. Go to Editor > Single Product template
2. Try to insert Product Price and Product Rating in couple of places
3. **Expected:** They should be available in the inserter

#### Other templates

1. Go to Editor > Product Catalog template
2. Try to insert Product Price and Product Rating within Product Collection block
3. **Expected:** They should be available in the inserter
4. Try to insert Product Price and Product Rating outside of Product Collection block
3. **Expected:** They should NOT be available in the inserter

#### Posts/Pages

1. Create new post
2. Insert Single Product block
3. Try to insert Product Price and Product Rating within Single Product block
4. **Expected:** They should be available in the inserter
5. Try to insert Product Price and Product Rating outside of Single Product block
3. **Expected:** They should NOT be available in the inserter

#### Patterns

1. Go to Editor > Patterns > Add new pattern
2. Create synced pattern
3. Insert Single Product block
4. Try to insert Product Price and Product Rating within Single Product block
5. **Expected:** They should be available in the inserter
6. Try to insert Product Price and Product Rating outside of Single Product block
3. **Expected:** They should NOT be available in the inserter

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/tests/e2e/tests/patterns/utils.ts
Comment (gigitux) [inline plugins/woocommerce-blocks/tests/e2e/tests/patterns/utils.ts:1]: This approach works for me, but I wonder if we should create a page object to standardize the way we write utilities across our E2E test suite.

File: plugins/woocommerce-blocks/tests/e2e/tests/related-products/related-products.block_theme.spec.ts
Comment (gigitux) [inline plugins/woocommerce-blocks/tests/e2e/tests/related-products/related-products.block_theme.spec.ts:1]: Are these changes related to this PR?

File: plugins/woocommerce-blocks/tests/e2e/tests/single-product-details/single-product-details.block_theme.spec.ts
Comment (gigitux) [inline plugins/woocommerce-blocks/tests/e2e/tests/single-product-details/single-product-details.block_theme.spec.ts:1]: Are these changes related to this PR?

**Review by gigitux — APPROVED** [review]
Awesome work! Thanks for working on this! 🙇 

I left some minor comments, but they are not blocker for merging this PR!

File: plugins/woocommerce-blocks/tests/e2e/tests/related-products/related-products.block_theme.spec.ts
Comment (kmanijak) [reply plugins/woocommerce-blocks/tests/e2e/tests/related-products/related-products.block_theme.spec.ts:1]: Yes. Previously it was expected the block is not registered hence trying to insert it threw an error. Now the block is registered so no error but it cannot be inserted as it's not available in this place.

File: plugins/woocommerce-blocks/tests/e2e/tests/single-product-details/single-product-details.block_theme.spec.ts
Comment (kmanijak) [reply plugins/woocommerce-blocks/tests/e2e/tests/single-product-details/single-product-details.block_theme.spec.ts:1]: Yes. Previously it was expected the block is not registered hence trying to insert it threw an error. Now the block is registered so no error but it cannot be inserted as it's not available in this place.

File: plugins/woocommerce-blocks/tests/e2e/tests/related-products/related-products.block_theme.spec.ts
Comment (gigitux) [reply plugins/woocommerce-blocks/tests/e2e/tests/related-products/related-products.block_theme.spec.ts:1]: Aaah nice! ❤️

File: plugins/woocommerce-blocks/tests/e2e/tests/single-product-details/single-product-details.block_theme.spec.ts
Comment (gigitux) [reply plugins/woocommerce-blocks/tests/e2e/tests/single-product-details/single-product-details.block_theme.spec.ts:1]: Aaah nice! ❤️

**Comment by kmanijak** [comment]
Merging as per https://github.com/woocommerce/woocommerce/issues/53078#issuecomment-2494089573
